### PR TITLE
DSL2: library merge

### PR DESCRIPTION
**Pull request by TCLamnidis**
Add Library merging SWF after Deduplication/Damage manipulation

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](<https://github.com/>nf-core/eager/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/eager _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint .`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by TCLamnidis** [comment]
I'm guessing linting fails have to do with the new template? I don't get all this red locally with nf-core tools 2.10

**Comment by jfy133** [comment]
Correct, go do that template merge boy!

File: subworkflows/local/merge_libraries.nf [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+workflow MERGE_LIBRARIES {
+    take:
+    ch_bam_bai  // [ [ meta ], bam , bai ]
+
+    main:
+    ch_versions      = Channel.empty()
+    ch_multiqc_files = Channel.empty()
+
+    ch_library_merge_input = ch_bam_bai
+        // TODO add 'id_index' to final meta? (once that also gets added to bam input). Maybe also keep SE/PE? (for now, we assume SE, see comment below)
```
Comment (jfy133) [inline subworkflows/local/merge_libraries.nf:19]: Still necessary?

File: subworkflows/local/merge_libraries.nf [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        .map { WorkflowEager.addNewMetaFromAttributes( it, ["id", "sample_id", "strandedness", "reference"], ["id", "sample_id", "strandedness", "reference"], false ) }
+        .groupTuple(by: 0)
+        // Discrad library-level metas, and bais. Add single_end: true to all metas (no SE/PE distinction at this point, right?)
+        .map {
+            meta, lib_metas, bam, bai ->
+            [ meta + [ 'single_end':true ], bam ]
```
Comment (jfy133) [inline subworkflows/local/merge_libraries.nf:25]:
```suggestion
            [ meta + [ 'single_end': true ], bam ]
```

File: workflows/eager.nf [hunk 3]
```diff
@@ -530,11 +531,22 @@ workflow EAGER {
         MANIPULATE_DAMAGE( ch_dedupped_bams, ch_fasta_for_deduplication.fasta, REFERENCE_INDEXING.out.pmd_masking )
         ch_multiqc_files       = ch_multiqc_files.mix( MANIPULATE_DAMAGE.out.flagstat.collect{it[1]}.ifEmpty([]) )
         ch_versions            = ch_versions.mix( MANIPULATE_DAMAGE.out.versions )
-        ch_bams_for_genotyping = params.genotyping_source == 'rescaled' ? MANIPULATE_DAMAGE.out.rescaled : params.genotyping_source == 'pmd' ? MANIPULATE_DAMAGE.out.filtered : params.genotyping_source == 'trimmed' ? MANIPULATE_DAMAGE.out.trimmed : ch_dedupped_bams
+        ch_bams_for_library_merge = params.genotyping_source == 'rescaled' ? MANIPULATE_DAMAGE.out.rescaled : params.genotyping_source == 'pmd' ? MANIPULATE_DAMAGE.out.filtered : params.genotyping_source == 'trimmed' ? MANIPULATE_DAMAGE.out.trimmed : ch_dedupped_bams
```
Comment (jfy133) [inline workflows/eager.nf:534]: Don't we need a branch here, if not multiple libraries then should skip?

**Review by jfy133 — APPROVED** [review]
Provisionally OK, but I think you're missing the single library skip re-routing

File: workflows/eager.nf (hunk 3 above)
Comment (TCLamnidis) [reply workflows/eager.nf:534]: running all samples through has the advantage that we get all final bams published in one directory. Would be nice to keep that feature even if skipping merging. Will look into how.

File: subworkflows/local/merge_libraries.nf (hunk 1 above)
Comment (TCLamnidis) [reply subworkflows/local/merge_libraries.nf:19]: BAM input would ideally get an `id_index`, just like FastQs do. I think the SE/PE bit doesn't really matter at this point, so hard-coding should be ok. I made an issue about the metas of BAM inputs #1040 
Deleteing the comment, though we should revisit this once bams have consistent meta info.

**Pull request merged**